src/dmd/staticcond.d

```diff
@@ -35,82 +37,391 @@ import dmd.utils;
+        if (e.isBool(true))
+            return true;
+        else if (e.isBool(false))
+        {
+            if (negatives)
+                negatives.push(before);
+            return false;
+        }
 
-    if (nerrors != global.errors ||
-        e.op == TOK.error ||
-        e.type.toBasetype() == Type.terror)
-    {
+        e.error("expression `%s` is not constant", e.toChars());
         errors = true;
         return false;
     }
+    return impl(e);
+}
+
+/********************************************
+ * Format a static condition as a tree-like structure, marking failed and
+ * bypassed expressions.
+ * Params:
+ *      original = original expression
+ *      instantiated = instantiated expression
+ *      negatives = array with negative clauses from `instantiated` expression
+ *      full = controls whether it shows the full output or only failed parts
+ *      itemCount = returns the number of written clauses
+ * Returns:
+ *      formatted string or `null` if the expressions were `null`
+ */
+const(char)* visualizeStaticCondition(Expression original, Expression instantiated,
+    const Expression[] negatives, bool full, ref uint itemCount)
+{
+    if (!original || !instantiated)
+        return null;
+
+    OutBuffer buf;
 
-    e = resolveAliasThis(sc, e);
+    if (full)
+        itemCount = visualizeFull(original, instantiated, negatives, buf);
+    else
+        itemCount = visualizeShort(original, instantiated, negatives, buf);
 
-    if (!e.type.isBoolean())
+    return buf.extractChars();
+}
+
+private uint visualizeFull(Expression original, Expression instantiated,
+    const Expression[] negatives, ref OutBuffer buf)
+{
+    // tree-like structure; traverse and format simultaneously
+    uint count;
+    uint indent;
+
+    static void printOr(uint indent, ref OutBuffer buf)
     {
-        exp.error("expression `%s` of type `%s` does not have a boolean value", exp.toChars(), e.type.toChars());
-        errors = true;
-        return false;
+        buf.reserve(indent * 4 + 8);
+        foreach (i; 0 .. indent)
+            buf.writestring("    ");
+        buf.writestring("    or:\n");
     }
 
-    e = e.ctfeInterpret();
+    // returns true if satisfied
+    bool impl(Expression orig, Expression e, bool inverted, bool orOperand, bool unreached)
+    {
+        TOK op = orig.op;
 
-    if (e.isBool(true))
-        return true;
-    else if (e.isBool(false))
-        return false;
+        // lower all 'not' to the bottom
+        // !(A && B) -> !A || !B
+        // !(A || B) -> !A && !B
+        if (inverted)
+        {
+            if (op == TOK.andAnd)
+                op = TOK.orOr;
+            else if (op == TOK.orOr)
+                op = TOK.andAnd;
+        }
 
-    e.error("expression `%s` is not constant", e.toChars());
-    errors = true;
-    return false;
+        if (op == TOK.not)
+        {
+            NotExp no = cast(NotExp)orig;
+            NotExp ne = cast(NotExp)e;
```

Comment:
This is assuming that `orig` and `e` have an identical structure. Is this invariant true? How does `e` differ, then?

Reply:
Now I check `.loc` of the toplevel expressions first, so they are either equal or syntax copied. Then I check for null with assert after these casts.